### PR TITLE
feat(gatsby): Add activity for writing out page-data.json files to the public directory

### PR DESCRIPTION
**Pull request by KyleAMathews**
This can take considerable time at scale (~18s for 100k pages) so we should let users know
what's going on + track it as there's optimizations we can add in the future.

Screen Shot 2021-06-18 at 10 09 08 AM

### CODE REVIEW

File: packages/gatsby/src/utils/page-data.ts [hunk 1]
```diff
@@ -136,7 +136,19 @@ export async function flush(): Promise<void> {
 
   const { pagePaths } = pendingPageDataWrites
 
-  const pagesToWrite = pagePaths.values()
+  const pagesToWrite = Array.from(pagePaths.values())
+
+  let writePageDataActivity
+  // Don't create the activity in tests as it changes
+  // behavior for the get-page-data tests.
```
Comment (KyleAMathews) [inline packages/gatsby/src/utils/page-data.ts:143]: having an activity changed the number of `setTimeouts` in https://github.com/gatsbyjs/gatsby/blob/589051f181e9b94f57f60c1eb3a5c4beb7425b82/packages/gatsby/src/utils/__tests__/get-page-data.ts#L207 I'm guessing the `.tick()` did that? Anyways, didn't seem worth changing so just turned off the activity during tests.

Comment (wardpeet) [reply packages/gatsby/src/utils/page-data.ts:143]: Please mock the createProgress reporter so we still test these pieces.

File: packages/gatsby/src/utils/page-data.ts [hunk 2]
```diff
@@ -136,7 +136,19 @@ export async function flush(): Promise<void> {
 
   const { pagePaths } = pendingPageDataWrites
 
-  const pagesToWrite = pagePaths.values()
+  const pagesToWrite = Array.from(pagePaths.values())
```
Comment (wardpeet) [inline packages/gatsby/src/utils/page-data.ts:139]: Let's change it to this. It's a Set that already only holds values.

```js
const { pagePaths: pagesToWrite } = pendingPageDataWrites
```

File: packages/gatsby/src/utils/page-data.ts [hunk 3]
```diff
@@ -136,7 +136,19 @@ export async function flush(): Promise<void> {
...
+
+  let writePageDataActivity
+  // Don't create the activity in tests as it changes
+  // behavior for the get-page-data tests.
+  if (process.env.NODE_ENV !== `test`) {
+    writePageDataActivity = reporter.createProgress(
+      `Writing page-data.json files to public directory`,
+      pagesToWrite.length,
```
Comment (wardpeet) [inline packages/gatsby/src/utils/page-data.ts:147]:
```suggestion
      pagesToWrite.size,
```

**Review by wardpeet — COMMENTED** [review]
Added a few comments as there is unnecessary code

File: packages/gatsby/src/utils/page-data.ts (hunk 2 above)
Comment (KyleAMathews) [reply packages/gatsby/src/utils/page-data.ts:139]: ah ok, typescript lied to me and just said it was an "IterableIterator" :-D

Comment (KyleAMathews) [reply packages/gatsby/src/utils/page-data.ts]: 👍

File: packages/gatsby/src/utils/page-data.ts (hunk 1 above)
Comment (KyleAMathews) [reply packages/gatsby/src/utils/page-data.ts]: 👍

**Review by wardpeet — APPROVED** [review]
Thanks! LGTM!

**Comment by LekoArts** [comment]
Published in `gatsby@3.8.1`

**Pull request merged**